### PR TITLE
[ble] Support multiple client connections (only up to 4 due to memory constraints)

### LINKED ISSUE #984: BLE needs to support multiple client connections
Currently it only allows one client to connect to it, and once the first client connects, the second client won't be able to connect.

### PR DESCRIPTION
**Pull request by jimmy-huang**
Refactored BLE module to handle multiple connections,
Zephyr default to only 1 client connection, so added a config
to overwrite it to support up to 4 clients.

Also enabled CONFIG_BLUETOOTH_DEBUG_LOG in debug mode to get more
more useful output.

Fixes #984, #334

Signed-off-by: Jimmy Huang <jimmy.huang@intel.com>

### CODE REVIEW

**Comment by grgustaf** [comment]
@jimmy-huang, if you rebase maybe this will work because we removed sensor from ashell?

File: src/zjs_ble.c [hunk 1]
```diff
@@ -375,14 +422,17 @@ static ZJS_DECL_FUNC(zjs_ble_update_value_callback_function)
...
-        if (jerry_get_object_native_handle(this, &ptr)) {
-            ble_characteristic_t *chrc = (ble_characteristic_t *)ptr;
-            if (chrc->chrc_attr) {
-                zjs_buffer_t *buf = zjs_buffer_find(argv[0]);
-                bt_gatt_notify(ble_conn.default_conn, chrc->chrc_attr,
+    uintptr_t ptr;
+    if (jerry_get_object_native_handle(this, &ptr)) {
+        ble_characteristic_t *chrc = (ble_characteristic_t *)ptr;
+        if (chrc->chrc_attr) {
+            zjs_buffer_t *buf = zjs_buffer_find(argv[0]);
```
Comment (grgustaf) [inline src/zjs_ble.c]: You may be, but just to be sure - you're guaranteed this will succeed, right?

File: src/zjs_ble.c [hunk 2]
```diff
@@ -464,45 +514,71 @@ static void zjs_ble_blvl_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16
 
 static void zjs_ble_connected_c_callback(void *handle, const void *argv)
 {
+    ble_handle_t * h = (ble_handle_t *)handle;
```
Comment (grgustaf) [inline src/zjs_ble.c]: drop space after first *, also twice below

File: src/zjs_ble.c [hunk 3]
```diff
@@ -464,45 +514,71 @@ static void zjs_ble_blvl_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16
...
 
+static void zjs_ble_disconnected_c_callback(void *handle, const void *argv)
+{
+    ble_handle_t * h = (ble_handle_t *)handle;
+    if (!h) {
+        ERR_PRINT("handle not found\n");
+        return;
+    }
+
+    char *addr = (char *)argv;
```
Comment (grgustaf) [inline src/zjs_ble.c]: You shouldn't be casting away const-ness like this. Can you fix that?

File: src/zjs_ble.c [hunk 4]
```diff
@@ -464,45 +514,71 @@ static void zjs_ble_blvl_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16
...
 {
     if (err) {
-        DBG_PRINT("Connection failed (err %u)\n", err);
+        ERR_PRINT("Connection failed (err %u)\n", err);
     } else {
-        DBG_PRINT("========== connected ==========\n");
+        DBG_PRINT("new connection\n");
+        ble_connection_t *new_conn = zjs_ble_new_connection(&ble_handle->connections, conn);
+        if (!new_conn) {
+            ERR_PRINT("failed to create new cilent connection\n");
```
Comment (grgustaf) [inline src/zjs_ble.c]: client

File: src/zjs_ble.c (hunk 1 above)
Comment (jimmy-huang) [inline src/zjs_ble.c]: what do you mean? you are saying I don't have to to check buffer? it will always exists?

Comment (grgustaf) [inline src/zjs_ble.c]: You're *not* checking it so I want to make sure you know it's always going to exist. That means you will never call this on an object that is not a valid buffer.

Comment (grgustaf) [inline src/zjs_ble.c]: Looking into it more myself, no it looks like this is directly called by JS so you can't be sure what they'll pass in. So you need to check whether buf is NULL here.

Comment (jimmy-huang) [inline src/zjs_ble.c]: @grgustaf added check now, i was looking at the read function before so I thought I already check the buffer

**Comment by grgustaf** [comment]
+1

**Pull request merged**